File: twolock.py
"""
twolock.py — the SECURITY GOAL: two locks ("Option B"). RESEARCH ARTIFACT.

THE WHOLE POINT OF THE PROJECT
------------------------------
The chaos keystream is UNVETTED. We never trust it on its own. So we never deploy it on its own.
Instead we wrap the data in TWO independent locks, one inside the other:

      plaintext  --[ INNER vault: AES-256-GCM ]-->  --[ OUTER wall: chaos AEAD ]-->  on the wire

  * INNER vault  = a real, world-standard, peer-reviewed cipher (AES-256-GCM, or
                   ChaCha20-Poly1305). This is the lock that actually GUARANTEES the data.
  * OUTER wall   = our chaos AEAD (aead.py). This is the exposed, sacrificial extra barrier that
                   an attacker hits FIRST and has to defeat before they even reach the vault.

THE GUARANTEE (the honest, important part)
------------------------------------------
The attacker on the wire sees only the outer chaos ciphertext. To get anywhere they must first
break the chaos wall. Suppose they do — suppose the chaos cipher is COMPLETELY broken tomorrow and
the attacker peels the outer layer off entirely. What's underneath is still AES-256-GCM. They are
exactly where they'd be against AES alone: facing a ~2^128 wall, with the data fully protected.

So the confidentiality and integrity of the plaintext rest on the VETTED inner vault. The chaos
layer can fail outright and the client loses nothing. That is why an unvetted cipher is safe to
ship HERE and nowhere else: it is never the only lock, and never the lock that matters.

This does NOT make the chaos math secure. It makes the DEPLOYMENT safe in spite of the chaos math
being unproven. The chaos wall buys: (1) an extra, real barrier an attacker must also break, and
(2) the exposed surface that gets battle-tested by real-world attacks — exactly what an unvetted
design needs to earn (or lose) trust over time.

WHY THIS ORDER (vetted INSIDE, chaos OUTSIDE) — not the reverse
---------------------------------------------------------------
If we put chaos on the inside and the vetted cipher outside, the thing protecting the actual
plaintext would be the chaos layer — so a chaos break would expose the plaintext, and the vetted
cipher would only protect an already-broken blob. Useless. The lock that the plaintext ultimately
depends on MUST be the vetted one, so the vetted cipher is the INNER vault. (Bonus: chaos on the
outside is the part exposed to attackers, which is where we want our unvetted experiment to live.)

KEY SEPARATION
--------------
The two locks NEVER share a key. From the caller's master key we derive two independent keys with
HKDF-SHA256 (a vetted KDF) under distinct labels — one for the inner vault, one for the outer wall.
A weakness or related-key effect in one lock cannot reach the other, and recovering the outer key
(e.g. by breaking chaos) reveals nothing about the inner key (HKDF is one-way).

Interface mirrors the other shells:

    blob = seal_twolock(master_key, plaintext, aad=b"", inner="aes-256-gcm")
    plaintext = open_twolock(master_key, blob, aad=b"")   # raises InvalidTag on tamper/wrong key

Wire format: the blob IS the outer chaos AEAD blob. What the chaos layer encrypts (its plaintext)
is:  alg(1) || inner_nonce(12) || inner_ciphertext+tag(N)
The alg byte (which vetted cipher is inside) rides INSIDE the authenticated+encrypted outer layer,
so an attacker can neither read nor change it, and open_twolock is self-describing — the caller does
not have to remember which inner cipher was used.

STILL UNVETTED (the chaos half). But the data's security here rests on the vetted half. See
THREAT_MODEL.md ("two locks") and REPORT.md.
"""

from __future__ import annotations

import os

from cryptography.hazmat.primitives import hashes
from cryptography.hazmat.primitives.ciphers.aead import AESGCM, ChaCha20Poly1305
from cryptography.hazmat.primitives.kdf.hkdf import HKDF
from cryptography.exceptions import InvalidTag as _CryptoInvalidTag

import aead as _outer

# Re-export so callers catch ONE exception type regardless of which lock rejected.
InvalidTag = _outer.InvalidTag

_HKDF_INNER_INFO = b"chaos-pwlcm-v1|twolock|inner-vault|v1"
_HKDF_OUTER_INFO = b"chaos-pwlcm-v1|twolock|outer-wall|v1"

INNER_NONCE_LEN = 12          # both AES-GCM and ChaCha20-Poly1305 take a 96-bit nonce
_KEY_LEN = 32                 # AES-256 / ChaCha20 key, and the outer master key

# alg id -> (name, AEAD class). The id is authenticated inside the outer layer.
_AES = 0x01
_CHACHA = 0x02
_INNER_BY_NAME = {"aes-256-gcm": _AES, "chacha20-poly1305": _CHACHA}
_INNER_BY_ID = {_AES: AESGCM, _CHACHA: ChaCha20Poly1305}


def _derive_keys(master_key: bytes) -> tuple[bytes, bytes]:
    """Split the caller's master key into two independent 32-byte keys via HKDF-SHA256 — one for the
    outer chaos wall, one for the inner vetted vault. Distinct info labels keep them domain-separated;
    HKDF's one-wayness means leaking one does not leak the other."""
    if not isinstance(master_key, (bytes, bytearray)):
        raise TypeError("master_key must be bytes")
    km = bytes(master_key)
    k_outer = HKDF(hashes.SHA256(), _KEY_LEN, None, _HKDF_OUTER_INFO).derive(km)
    k_inner = HKDF(hashes.SHA256(), _KEY_LEN, None, _HKDF_INNER_INFO).derive(km)
    return k_outer, k_inner


def seal_twolock(master_key: bytes, plaintext: bytes, aad: bytes = b"",
                 inner: str = "aes-256-gcm", *,
                 inner_nonce: bytes | None = None, outer_nonce: bytes | None = None) -> bytes:
    """Encrypt under two independent locks: the vetted inner vault, then the chaos outer wall.

    `aad` is bound to BOTH locks (defense in depth — both layers authenticate the context). `inner`
    selects the vetted cipher ("aes-256-gcm" default, or "chacha20-poly1305"). Returns the outer
    chaos blob; each call uses fresh random nonces in both layers, so the same plaintext seals
    differently every time.

    `inner_nonce` / `outer_nonce` are keyword-only and exist ONLY to pin the two sources of
    nondeterminism for a known-answer test / Rust parity vector (mirrors aead.seal's `nonce=` and
    streaming.seal_stream's `salt=`). Leave them None in real use — fresh random nonces are the safe
    default; reusing a nonce across two different plaintexts breaks the vetted vault."""
    try:
        alg = _INNER_BY_NAME[inner]
    except KeyError:
        raise ValueError(f"unknown inner cipher {inner!r}; "
                         f"choose one of {sorted(_INNER_BY_NAME)}") from None

    k_outer, k_inner = _derive_keys(master_key)

    # INNER vault: real, vetted AEAD with its own fresh nonce (pinned only for a KAT).
    if inner_nonce is None:
        inner_nonce = os.urandom(INNER_NONCE_LEN)
    elif len(inner_nonce) != INNER_NONCE_LEN:
        raise ValueError(f"inner_nonce must be exactly {INNER_NONCE_LEN} bytes")
    inner_ct = _INNER_BY_ID[alg](k_inner).encrypt(inner_nonce, plaintext, aad)
    inner_blob = bytes([alg]) + inner_nonce + inner_ct

    # OUTER wall: our chaos AEAD wraps the whole inner blob (and binds the same aad again). The outer
    # nonce, when pinned, is forwarded to aead.seal's own KAT hook; None there means a fresh random one.
    return _outer.seal(k_outer, inner_blob, aad=aad, nonce=outer_nonce)


def open_twolock(master_key: bytes, blob: bytes, aad: bytes = b"") -> bytes:
    """Peel the outer chaos wall, then open the inner vetted vault. Raises InvalidTag if EITHER lock
    rejects — wrong key or tampering at any layer — and NEVER returns plaintext in that case. The
    inner cipher is read from the (authenticated) blob, so the caller need not specify it."""
    k_outer, k_inner = _derive_keys(master_key)

    # OUTER wall: chaos AEAD verifies + decrypts; raises InvalidTag on tamper/wrong outer key.
    inner_blob = _outer.open_(k_outer, blob, aad=aad)
    if len(inner_blob) < 1 + INNER_NONCE_LEN:
        raise InvalidTag("inner blob too short / malformed")

    alg = inner_blob[0]
    cls = _INNER_BY_ID.get(alg)
    if cls is None:
        raise InvalidTag(f"unknown inner cipher id {alg:#x}")
    inner_nonce = inner_blob[1:1 + INNER_NONCE_LEN]
    inner_ct = inner_blob[1 + INNER_NONCE_LEN:]

    # INNER vault: the lock that actually guarantees the data. Even if the outer chaos wall were
    # fully broken, an attacker would still be stopped right here by the vetted cipher.
    try:
        return cls(k_inner).decrypt(inner_nonce, inner_ct, aad)
    except _CryptoInvalidTag:
        raise InvalidTag("inner vault authentication failed — wrong key or tampered ciphertext") from None


if __name__ == "__main__":
    key = b"my shared secret key (any bytes)"
    msg = b"Attack at dawn. Wire $40,000 by Friday."

    blob = seal_twolock(key, msg)
    print(f"sealed ({len(blob)} bytes): {blob.hex()[:48]}...")
    print(f"opened: {open_twolock(key, blob)!r}")

    # ChaCha20-Poly1305 inner works too, and the blob is self-describing on open.
    cblob = seal_twolock(key, msg, inner="chacha20-poly1305")
    print(f"chacha inner opened: {open_twolock(key, cblob)!r}")

    # Same message twice -> different blobs (fresh nonces in both layers).
    print(f"two seals differ: {seal_twolock(key, msg) != seal_twolock(key, msg)}")

    # Tamper one byte -> rejected by the outer wall.
    bad = bytearray(blob)
    bad[-1] ^= 0x01
    try:
        open_twolock(key, bytes(bad))
        print("TAMPER NOT DETECTED  <-- BUG")
    except InvalidTag as e:
        print(f"tamper rejected: {e}")

    # Wrong key -> rejected.
    try:
        open_twolock(b"the wrong key................", blob)
        print("WRONG KEY ACCEPTED  <-- BUG")
    except InvalidTag as e:
        print(f"wrong key rejected: {e}")

File: rust/src/twolock.rs
//! Two-locks wrapper — chaos OUTER wall over a VETTED inner vault (Phase 8.4).
//!
//! Port of twolock.py. THE SECURITY GOAL of the whole project. The chaos keystream
//! is UNVETTED, so it is NEVER the only lock. We wrap the data in two independent
//! locks, one inside the other:
//!
//!    plaintext --[ INNER vault: AES-256-GCM / ChaCha20-Poly1305 ]-->
//!              --[ OUTER wall: chaos AEAD ]--> wire
//!
//! Even a TOTAL chaos break leaves the attacker facing AES-256-GCM (~2^128),
//! data intact. That is why an unvetted cipher is safe to ship HERE and nowhere else.
//!
//! STILL UNVETTED (the chaos half). But the data's security rests on the vetted half.

use aes_gcm::aead::{generic_array::GenericArray, Aead, KeyInit, Payload};
use aes_gcm::Aes256Gcm;
use chacha20poly1305::ChaCha20Poly1305;
use hkdf::Hkdf;
use sha2::Sha256;
use zeroize::Zeroizing;

use crate::aead::{aead_open, aead_seal};

pub const INNER_NONCE_LEN: usize = 12;
const TWOLOCK_KEY_LEN: usize = 32;

/// Inner-cipher selector. Authenticated+encrypted inside the outer layer.
pub const TWOLOCK_AES: u8 = 0x01;
pub const TWOLOCK_CHACHA: u8 = 0x02;

const TWOLOCK_HKDF_INNER_INFO: &[u8] = b"chaos-pwlcm-v1|twolock|inner-vault|v1";
const TWOLOCK_HKDF_OUTER_INFO: &[u8] = b"chaos-pwlcm-v1|twolock|outer-wall|v1";

/// One HKDF-SHA256 derivation (salt=None, matching Python's HKDF).
fn hkdf_sha256_key(master_key: &[u8], info: &[u8]) -> Zeroizing<[u8; TWOLOCK_KEY_LEN]> {
    let hk = Hkdf::<Sha256>::new(None, master_key);
    let mut okm = Zeroizing::new([0u8; TWOLOCK_KEY_LEN]);
    hk.expand(info, &mut okm[..])
        .expect("32 bytes is a valid HKDF-SHA256 output length");
    okm
}

/// Split the master key into (outer-wall key, inner-vault key).
fn twolock_derive_keys(
    master_key: &[u8],
) -> (
    Zeroizing<[u8; TWOLOCK_KEY_LEN]>,
    Zeroizing<[u8; TWOLOCK_KEY_LEN]>,
) {
    let k_outer = hkdf_sha256_key(master_key, TWOLOCK_HKDF_OUTER_INFO);
    let k_inner = hkdf_sha256_key(master_key, TWOLOCK_HKDF_INNER_INFO);
    (k_outer, k_inner)
}

/// Generic vetted-AEAD encrypt.
fn vault_seal<C: Aead + KeyInit>(
    key: &[u8],
    nonce: &[u8],
    pt: &[u8],
    aad: &[u8],
) -> Option<Vec<u8>> {
    let cipher = C::new_from_slice(key).ok()?;
    cipher
        .encrypt(GenericArray::from_slice(nonce), Payload { msg: pt, aad })
        .ok()
}

/// Generic vetted-AEAD decrypt.
fn vault_open<C: Aead + KeyInit>(
    key: &[u8],
    nonce: &[u8],
    ct: &[u8],
    aad: &[u8],
) -> Option<Vec<u8>> {
    let cipher = C::new_from_slice(key).ok()?;
    cipher
        .decrypt(GenericArray::from_slice(nonce), Payload { msg: ct, aad })
        .ok()
}

fn inner_vault_seal(
    alg: u8,
    key: &[u8],
    nonce: &[u8],
    pt: &[u8],
    aad: &[u8],
) -> Option<Vec<u8>> {
    match alg {
        TWOLOCK_AES => vault_seal::<Aes256Gcm>(key, nonce, pt, aad),
        TWOLOCK_CHACHA => vault_seal::<ChaCha20Poly1305>(key, nonce, pt, aad),
        _ => None,
    }
}

fn inner_vault_open(
    alg: u8,
    key: &[u8],
    nonce: &[u8],
    ct: &[u8],
    aad: &[u8],
) -> Option<Vec<u8>> {
    match alg {
        TWOLOCK_AES => vault_open::<Aes256Gcm>(key, nonce, ct, aad),
        TWOLOCK_CHACHA => vault_open::<ChaCha20Poly1305>(key, nonce, ct, aad),
        _ => None,
    }
}

/// Seal under two independent locks: the vetted inner vault, then the chaos outer wall.
pub fn twolock_seal(
    master_key: &[u8],
    outer_nonce: &[u8],
    inner_nonce: &[u8],
    plaintext: &[u8],
    aad: &[u8],
    inner_alg: u8,
    n_maps: usize,
) -> Option<Vec<u8>> {
    let (k_outer, k_inner) = twolock_derive_keys(master_key);
    let inner_ct =
        inner_vault_seal(inner_alg, &k_inner[..], &inner_nonce[..INNER_NONCE_LEN], plaintext, aad)?;
    let mut inner_blob = Vec::with_capacity(1 + INNER_NONCE_LEN + inner_ct.len());
    inner_blob.push(inner_alg);
    inner_blob.extend_from_slice(&inner_nonce[..INNER_NONCE_LEN]);
    inner_blob.extend_from_slice(&inner_ct);
    Some(aead_seal(
        &k_outer[..],
        outer_nonce,
        &inner_blob,
        aad,
        n_maps,
    ))
}

/// Peel the chaos outer wall, then open the vetted inner vault.
pub fn twolock_open(
    master_key: &[u8],
    blob: &[u8],
    aad: &[u8],
    n_maps: usize,
) -> Option<Vec<u8>> {
    let (k_outer, k_inner) = twolock_derive_keys(master_key);
    let inner_blob = aead_open(&k_outer[..], blob, aad, n_maps)?;
    if inner_blob.len() < 1 + INNER_NONCE_LEN {
        return None;
    }
    let alg = inner_blob[0];
    let inner_nonce = &inner_blob[1..1 + INNER_NONCE_LEN];
    let inner_ct = &inner_blob[1 + INNER_NONCE_LEN..];
    inner_vault_open(alg, &k_inner[..], inner_nonce, inner_ct, aad)
}

#[cfg(test)]
mod tests {
    use super::*;
    use crate::aead::aead_open as chaos_aead_open;
    use crate::constants::DEFAULT_N_MAPS;

    #[test]
    fn twolock_roundtrip_both_inner_ciphers() {
        let master = b"two-locks master key";
        let outer_nonce = b"outer-nonce-16by";
        let inner_nonce = b"inner-nonce1";
        let aad = b"context";
        let pt = b"vault the secret behind two independent locks";
        for alg in [TWOLOCK_AES, TWOLOCK_CHACHA] {
            let blob = twolock_seal(master, outer_nonce, inner_nonce, pt, aad, alg, DEFAULT_N_MAPS)
                .unwrap();
            let got = twolock_open(master, &blob, aad, DEFAULT_N_MAPS).unwrap();
            assert_eq!(got, pt, "two-locks round trip failed for inner alg {alg:#x}");
        }
    }

    #[test]
    fn twolock_rejects_tamper_wrong_key_and_wrong_aad() {
        let master = b"two-locks master key";
        let outer_nonce = b"outer-nonce-16by";
        let inner_nonce = b"inner-nonce1";
        let aad = b"context";
        let pt = b"defense in depth";
        let blob =
            twolock_seal(master, outer_nonce, inner_nonce, pt, aad, TWOLOCK_AES, DEFAULT_N_MAPS)
                .unwrap();

        assert!(twolock_open(b"the wrong master key", &blob, aad, DEFAULT_N_MAPS).is_none());
        assert!(twolock_open(master, &blob, b"other aad", DEFAULT_N_MAPS).is_none());
        let mut bad = blob.clone();
        bad[crate::aead::NONCE_LEN + crate::aead::COMMIT_LEN + 1] ^= 0x01;
        assert!(twolock_open(master, &bad, aad, DEFAULT_N_MAPS).is_none());
    }

    #[test]
    fn twolock_key_separation_outer_key_does_not_open_inner() {
        let master = b"two-locks master key";
        let (k_outer, k_inner) = twolock_derive_keys(master);
        assert_ne!(&k_outer[..], &k_inner[..]);

        let blob = twolock_seal(
            master,
            b"outer-nonce-16by",
            b"inner-nonce1",
            b"top secret",
            b"a",
            TWOLOCK_AES,
            DEFAULT_N_MAPS,
        )
        .unwrap();
        let inner_blob = chaos_aead_open(&k_outer[..], &blob, b"a", DEFAULT_N_MAPS).unwrap();
        let alg = inner_blob[0];
        let nonce = &inner_blob[1..1 + INNER_NONCE_LEN];
        let ct = &inner_blob[1 + INNER_NONCE_LEN..];
        assert!(inner_vault_open(alg, &k_inner[..], nonce, ct, b"a").is_some());
        assert!(inner_vault_open(alg, &k_outer[..], nonce, ct, b"a").is_none());
    }
}
